queryable-spec.ts

```diff
@@ -0,0 +1,390 @@
+
+/*
+ * Helper union type 
+ */
+type TermName = 'subject' | 'predicate' | 'object' | 'graph';
+
+/*
+ * Custom typings for the RDF/JS Stream interface as the current
+ * typings restrict the generic param Q to extensions of "BaseQuad",
+ * meaning it cannot be used for Bindings.
+ */
+export interface Stream<Q> extends EventEmitter {
+  read(): Q | null;
+}
+
+
+/******************************************************************************
+                              FILTERABLE SOURCE
+ *****************************************************************************/
+
+
+/**
+ * QueryResultMetadataOptions is an abstract interface that represents a generic 
+ * expression over a stream of quads. 
+ */
+interface Expression {
+  /**
+   * Value that identifies the concrete interface of the expression, since the
+   * Expression itself is not directly instantiated. Possible values include 
+   * "operator" and "term".
+   */
+  expressionType: string;
+};
+
+/**
+ * An OperatorExpression is represents an expression that applies a given
+ * operator on given sub-expressions.
+ * 
+ * The WebIDL definition of the Filterable spec contains a list of supported 
+ * operators: https://rdf.js.org/query-spec/#expression-operators 
+ */
+interface OperatorExpression extends Expression {
+  
+  /** 
+   * Contains the constant "operator". 
+   */
+  expressionType: 'operator';
+
+  /**
+   * Value that identifies an operator. Possible values can be found in the 
+   * list of operators.
+   */
+  operator: string;
+
+  /**
+   * Array of Expression's on to which the given operator applies. The length
+   * of this array depends on the operator.
+   */
+  args: Expression[];
+};
+
+/**
+ * A TermExpression is an expression that contains a Term.
+ */
+interface TermExpression {
+  
+  /**
+   * The constant "term".
+   */
+  expressionType: 'term';
+
+  /**
+   * a Term.
+   */
+  term: RDF.Term;
+}
+
+/**
+ * ExpressionFactory enables expressions to be created in an idiomatic manner.
+ */
+interface ExpressionFactory {
+  
+  /**
+   * Creates a new OperatorExpression instance for the given operator and array of arguments.
+   */
+  operatorExpression(operator: string, args: Expression[]): OperatorExpression;
+
+  /**
+   * Creates a new TermExpression instance for the given term.
+   */
+  termExpression(term: RDF.Term): TermExpression;
+};
+
+
+/**
+ * QueryResultMetadataCount is part of the QueryResultMetadata interface to
+ * represent metadata about the number of quads in the result stream.
+ */
+interface FilterableResultMetadataCount {
+  /**
+   * indicates the type of counting that was done, and MUST either be 
+   * "estimate" or "exact".
+   */
+  type: 'estimate' | 'exact';
+
+  /**
+   * Indicates an estimate of the number of quads in the stream if 
+   * type = "estimate", or the exact number of quads in the stream if 
+   * type = "exact".
+   */
+  value: number;
+};
+
+/**
+ * TBD
+ */
+ interface FilterableCost {
```

Comment:
```suggestion
 interface FilterableCost extends Record<string, any> {
```
For extensibility?

Reply:
Perhaps we can name this `QueryOperationCost`, as it's probably reusable in other places?

Reply:
Any preference when it comes to `extends Record<string, any>` vs. `[key: string]: any` ?

Reply:
Ah no, no preference here :-)